drivers/tee/optee/optee.c

```diff
@@ -700,15 +720,38 @@ static void handle_rpc_call(const struct device *dev, struct optee_rpc_param *pa
 	param->a0 = res;
 }
 
-static int optee_call(const struct device *dev, struct optee_msg_arg *arg)
+static inline void *get_off_addr(void *addr, unsigned int offset)
+{
+	return ((uint8_t *)addr + offset);
+}
+
+static int optee_call(const struct device *dev, struct optee_msg_arg *arg, struct tee_shm *shm)
 {
 	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
-	struct optee_rpc_param param = {
-		.a0 = OPTEE_SMC_CALL_WITH_ARG
-	};
+	struct optee_rpc_param param = { };
+	struct optee_msg_arg *rpc_arg = NULL;
+	unsigned int rpc_arg_offs;
 	void *pages = NULL;
 
-	u64_to_regs((uint64_t)optee_mem_phys_addr(arg), &param.a1, &param.a2);
+	if (data->rpc_param_count != 0) {
+		rpc_arg_offs = OPTEE_MSG_GET_ARG_SIZE(arg->num_params);
+		rpc_arg = get_off_addr(shm->addr, rpc_arg_offs);
+	}
+
+	/* For the registered memory REGD_ARG should be sent */
+	if (rpc_arg && shm->flags & TEE_SHM_REGISTER) {
+		param.a0 = OPTEE_SMC_CALL_WITH_REGD_ARG;
+		u64_to_regs((uint64_t)shm, &param.a1, &param.a2);
+	} else {
+		if (rpc_arg != 0) {
+			param.a0 = OPTEE_SMC_CALL_WITH_RPC_ARG;
```

Comment:
AFAIK, this is a relatively new feature. Not all OP-TEE versions support it (I am looking at you, Renesas). 
So you can't use it unconditionally. You need to check for `OPTEE_SMC_SEC_CAP_RPC_ARG` before using it.

Reply:
It is checked indirectly through rpc_arg param, which depends on rpc_param_count, which depends on OPTEE_SMC_SEC_CAP_RPC_ARG. While it is a little convoluted I believe it's better to leave it like that to be more in tune with reference Linux implementation/